File: backend/app/models/catalog.py
from enum import StrEnum

from pydantic import BaseModel, Field


class ComponentCategory(StrEnum):
    CORE_EQUIPMENT = "core_equipment"
    ACCESSORY = "accessory"
    FEE = "fee"
    SERVICE = "service"
    MOUNTING = "mounting"
    PACKAGE = "package"
    OTHER = "other"


class ComponentKind(StrEnum):
    PV_MODULE = "pv_module"
    INVERTER = "inverter"
    BATTERY = "battery"
    HEAT_PUMP = "heat_pump"
    WALLBOX = "wallbox"
    ACCESSORY = "accessory"
    INSTALLATION_FEE = "installation_fee"
    SERVICE_FEE = "service_fee"
    MOUNTING = "mounting"
    PACKAGE = "package"
    OTHER = "other"


class ComponentSpec(BaseModel):
    module_watt_peak: float | None = None
    inverter_power_kw: float | None = None
    battery_capacity_kwh: float | None = None
    wallbox_charging_power_kw: float | None = None
    heatpump_nominal_power_kw: float | None = None


class CatalogComponent(BaseModel):
    id: str
    component_type: str
    component_name: str
    component_brand: str | None = None
    technology: str | None = None
    category: ComponentCategory
    kind: ComponentKind
    spec: ComponentSpec = Field(default_factory=ComponentSpec)
    source_datasets: list[str]
    observed_count: int
    warnings: list[str] = Field(default_factory=list)


class CatalogSummary(BaseModel):
    component_count: int
    source_datasets: list[str]
    counts_by_category: dict[str, int]
    counts_by_kind: dict[str, int]
    warning_count: int


class ComponentCatalog(BaseModel):
    summary: CatalogSummary
    components: list[CatalogComponent]

File: frontend/src/lib/catalog.ts
export type CatalogComponentSpec = {
  module_watt_peak?: number | null;
  inverter_power_kw?: number | null;
  battery_capacity_kwh?: number | null;
  wallbox_charging_power_kw?: number | null;
  heatpump_nominal_power_kw?: number | null;
};

export type CatalogComponent = {
  id: string;
  component_type: string;
  component_name: string;
  component_brand: string | null;
  technology: string | null;
  category: string;
  kind: string;
  spec: CatalogComponentSpec;
};

export type CatalogResponse = {
  summary: { component_count: number };
  components: CatalogComponent[];
};

export type PanelDimensions = {
  lengthMeters: number;
  widthMeters: number;
  thicknessMeters: number;
};

export const DEFAULT_PANEL: PanelDimensions = {
  lengthMeters: 1.762,
  widthMeters: 1.134,
  thicknessMeters: 0.04,
};

export function fetchPvModules(): Promise<CatalogComponent[]> {
  return fetch("/api/catalog?kind=pv_module")
    .then((r) => (r.ok ? (r.json() as Promise<CatalogResponse>) : null))
    .then((data) => data?.components ?? [])
    .catch(() => []);
}

export function pickPrimaryPanel(modules: CatalogComponent[]): CatalogComponent | null {
  if (!modules.length) return null;
  return (
    modules
      .filter((m) => (m.spec.module_watt_peak ?? 0) > 0)
      .sort(
        (a, b) => (b.spec.module_watt_peak ?? 0) - (a.spec.module_watt_peak ?? 0),
      )[0] ?? modules[0]
  );
}

export function dimensionsFor(_module: CatalogComponent | null): PanelDimensions {
  return DEFAULT_PANEL;
}
